---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

#load package
```{r}
#load all the library that needed
library(clusterProfiler)
library(DOSE)
library(org.Hs.eg.db)
library(enrichplot)
library(dplyr)
library(ggplot2)
library(stringr)
library(biomaRt)
library(tibble)
library(DESeq2)
library(limma)
library("FactoMineR")
library("factoextra")
library(ggrepel)
library(patchwork)
library(hrbrthemes)
library(rrvgo)
library(tximportData)
library(tximport)
library(scales)
library(igraph)
library(pheatmap)
library(RColorBrewer)
```

```{r}
sessionInfo()
```

#Functions
```{r}
#GO enrichment and KEGG pathway
GO_enrichment_human = function (a,b) {
  library(clusterProfiler)
  library(DOSE)
  library(org.Hs.eg.db)
  library(enrichplot)
  library(dplyr)
  enrichGO(gene = a$genes,
                keyType       = 'ENSEMBL',
                OrgDb         = org.Hs.eg.db,
                ont           = b,
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                minGSSize     = 5,
                readable      = TRUE)
}

KEGG_fun_human <- function (x) {
  library(clusterProfiler)
  library(DOSE)
  library(org.Hs.eg.db)
  library(enrichplot)
  library(dplyr)
  enrichKEGG(gene = (bitr(x$genes, fromType = "ENSEMBL", toType = c("ENTREZID"),OrgDb = org.Hs.eg.db)) $ENTREZID,
                 organism     = 'hsa',
                 pvalueCutoff = 0.05)
}


#PCA plot custome script to adpot my plotting better.
cus.plotPCA_34 <- function (object, intgroup = "condition", returnData = FALSE, ntop = Inf) #ntop is to define how many gene you take for consideration
{ library(ggplot2)
  library(dplyr)
  library(ggrepel)
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(PC3 = pca$x[, 3], PC4 = pca$x[, 4], group = group, 
                  intgroup.df, exp = colData(object)[,2], name=rownames(colData(object)))
  if (returnData) {
    attr(d, "percentVar") <- percentVar[3:4]
    return(d)
  }
    ggplot(data = d, aes_string(x = "PC3", y = "PC4", color = "group", shape="exp")) + geom_point(size = 8) + xlab(paste0("PC3: ", round(percentVar[3] * 100), "% variance")) + ylab(paste0("PC4: ", round(percentVar[4] * 100), "% variance")) + coord_fixed()  + theme_bw(base_size = 15) + theme(axis.text = element_text(size = 15)) + scale_color_manual(values=c("#FE920A","#7F7F7F"))
}

cus.plotPCA_45 <- function (object, intgroup = "condition", returnData = FALSE, ntop = Inf) 
{ library(ggplot2)
  library(dplyr)
  library(ggrepel)
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(PC4 = pca$x[, 4], PC5 = pca$x[, 5], group = group, 
                  intgroup.df, exp = colData(object)[,2], name=rownames(colData(object)))
  if (returnData) {
    attr(d, "percentVar") <- percentVar[4:5]
    return(d)
  }
    ggplot(data = d, aes_string(x = "PC4", y = "PC5", color = "group", shape="exp")) + geom_point(size = 8) + xlab(paste0("PC4: ", round(percentVar[4] * 100), "% variance")) + ylab(paste0("PC5: ", round(percentVar[5] * 100), "% variance")) + coord_fixed()  + theme_bw(base_size = 15) + theme(axis.text = element_text(size = 15)) + scale_color_manual(values=c("#FE920A","#7F7F7F"))
}

cus.plotPCA_12 <- function (object, intgroup = "condition", returnData = FALSE, ntop = Inf)  
{ library(ggplot2)
  library(dplyr)
  library(ggrepel)
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(PC1 = pca$x[, 1], PC2 = pca$x[, 2], group = group, 
                  intgroup.df, exp = colData(object)[,2], name=rownames(colData(object)))
  if (returnData) {
    attr(d, "percentVar") <- percentVar[1:2]
    return(d)
  }
    ggplot(data = d, aes_string(x = "PC1", y = "PC2", color = "group", shape="exp")) + geom_point(size = 8) + xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) + ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) + coord_fixed() +  theme_bw(base_size = 15) + theme(axis.text = element_text(size = 15)) + scale_color_manual(values=c("#FE920A","#7F7F7F")) 
}


#some small changes for plotting enriched GO terms cluster. y is the number plot, note that after                                                   
custome_emapplot = function (x,y, z, m, nCluster=NULL, node_label="category") {
  x_pair_t <- enrichplot::pairwise_termsim (x)
  x_pair_t@result$lgpadj <- -log10(x_pair_t@result$p.adjust)
  show_cate <- x@result %>% 
    dplyr::filter(p.adjust <= 0.05) %>%
    dplyr::slice_min(n=y, order_by=p.adjust)
  emapplot(x_pair_t, showCategory = show_cate$Description, group_category=T, group_legend=T, color="lgpadj", cex_category=z, cex_label_category=m, nCluster=nCluster, node_label=node_label) + scale_fill_gradient(high = "black", low = "white", name="-log10(p.adj)")
}

#prepare data for DE gene distribution
DEnumber_chrsize = function (x) {
  library(dplyr)  
  a <- x %>% 
    dplyr::count(chromosome_name)
    a$chromosome_name <- sub("^", "chr", a$chromosome_name)
    a <- dplyr::full_join(a, chr, by=c("chromosome_name"="V1"))
    a[is.na(a)] <- 0
    colnames(a) <- NULL
    colnames(a) <- c("chr", "DEnumber", "chr_size")
    a$DEnumber <- as.numeric(a$DEnumber)
    a$chr_size <- as.numeric(a$chr_size)
    a$DE_density <- a$DEnumber / a$chr_size
    a <- a %>%
      dplyr::filter(!chr=="chrMT") #filter out 2 DE genes on mitochondria
    print(a)
}

#to fix the number of axis showing
fancy_scientific <- function(l) {
     # turn in to character string in scientific notation
     l <- format(l, scientific = TRUE)
     # quote the part before the exponent to keep all the digits
     l <- gsub("^(.*)e", "'\\1'e", l)
     # turn the 'e+' into plotmath format
     l <- gsub("e", "%*%10^", l)
     # return this as an expression
     parse(text=l)
}

#x is th KEGG enrichment result, y is for printing, "UP" or "DOWN", "activated" or "supressed", selected top z based on count/Generatio/bgratio
get_KEGG_info_2 = function (x,y,z) {
  library(dplyr)
  x@result %>%
  dplyr::filter(p.adjust <= 0.05) %>%
  dplyr::slice_min(n=z, order_by=p.adjust) %>%
  dplyr::select(c(ID, Description, p.adjust, geneID, Count)) %>%
  dplyr::mutate(trend= y) %>%
  dplyr::arrange(desc(Count)) 
}

```


#DE analysis and PCA
```{r}
#import the readcount table (already clean up, clip the first row with all the command line, remove the annotation position, only keep ensemble gene ID, length and the count)
GSE147770_C123_S123_210723 <- read.delim("/Users/keyigeng/Dropbox/HAP1_Gen_insta/readscount/GSE147770_C123_S123_210723_clean.readCount") 
GSE103405_GSE148175_210723 <- read.delim("/Users/keyigeng/Dropbox/HAP1_Gen_insta/readscount/GSE103405_GSE148175_210723_clean.readCount")
RNA_seq_191215 <- read.delim("/Users/keyigeng/Dropbox/illuminadata/RNA_seq_191215/featurecount/HAP1_px459_635645_191217_genecode_clean.readCount")
RNA_seq_191215$Genes <- str_replace(RNA_seq_191215$Geneid,
                        pattern = ".[0-9]+$",
                        replacement = "")
#pack all the four dataset into one table
allfour_HAP1_RNAseq <- RNA_seq_191215 %>%
  dplyr::inner_join(GSE147770_C123_S123_210723, by=c("Geneid"="Geneid")) %>%
  dplyr::inner_join(GSE103405_GSE148175_210723, by=c("Geneid"="Geneid")) %>%
  dplyr::select(!c("Length.y", "Geneid")) 
colnames(allfour_HAP1_RNAseq) <- c("our_ctrl_r1", "our_ctrl_r2","our_td_r1", "our_td_r2","Genes", "length", "GSE147770_C1", "GSE147770_C2", "GSE147770_C3", "GSE147770_S1", "GSE147770_S2", "GSE147770_S3", "GSE103405_ko1", "GSE103405_ko2", "GSE103405_wt1", "GSE103405_wt2","GSE148175_koSMARCC1_r1", "GSE148175_koSMARCC1_r2", "GSE148175_koSMARCC2_r1", "GSE148175_koSMARCC2_r2")

#use DEseq2 to do differential expression analysis
#prepare input file format
allfour_HAP1_RNAseq_input <- allfour_HAP1_RNAseq %>%
  dplyr::select(!c("length")) %>%
  tibble::column_to_rownames(var="Genes")
condition_allfour <-c(rep(c("normal", "deletion"), each=2), rep(c("normal", "deletion"), each=3), rep(c("deletion", "normal"), each=2), rep(c("deletion", "normal"), each=2))
batch_allfour <- c(rep("ours", each=4), rep("GSE147770", each=6), rep("GSE103405", each=4), rep("GSE148175", each=4))
coldata_allfour <- data.frame (cbind(condition_allfour, batch_allfour))
rownames(coldata_allfour) <- colnames(allfour_HAP1_RNAseq_input)

dds_allfour_RNAseq <-DESeqDataSetFromMatrix((allfour_HAP1_RNAseq_input %>% as.matrix() %>% round()), 
                                                         colData= coldata_allfour,
                                                         design = ~batch_allfour + condition_allfour)
dds_allfour_RNAseq <- DESeq(dds_allfour_RNAseq)
all(rownames(coldata_allfour) %in% colnames(allfour_HAP1_RNAseq_input))
res_allfour_RNAseq <- results(dds_allfour_RNAseq, contrast = c("condition_allfour","deletion","normal"))
allfour_RNAseq_deseqnor_count <- counts(dds_allfour_RNAseq, normalized=T)

mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host="https://may2021.archive.ensembl.org") 
ensembl2name_loc <- getBM(attributes=c("ensembl_gene_id","external_gene_name","chromosome_name","start_position", "end_position"),mart=mart)
allfour_HAP1_RNAseq_DEdeseq2 <- data.frame(res_allfour_RNAseq[ which(res_allfour_RNAseq$padj <= 0.05 ),]) %>%
  add_rownames(var="genes") %>%
  dplyr::inner_join(ensembl2name_loc, by=c("genes"="ensembl_gene_id")) 

allfour_HAP1_RNAseq_deseq2 <- data.frame(res_allfour_RNAseq) %>%
  add_rownames(var="genes") %>%
  dplyr::inner_join(ensembl2name_loc, by=c("genes"="ensembl_gene_id")) %>%
  dplyr::select(!c(9,10,11))

#Prepare data for PCA plot.
#apply batch factor correction
dds_allfour_rl_bf <- rlog(dds_allfour_RNAseq, blind=FALSE)
dds_allfour_rl_bf_as <- assay(dds_allfour_rl_bf)
dds_allfour_rl_bf_as <- limma::removeBatchEffect(dds_allfour_rl_bf_as, dds_allfour_rl_bf$batch_allfour)
assay(dds_allfour_rl_bf) <- dds_allfour_rl_bf_as
#no batch effect corrected
dds_allfour_rl <- rlog(dds_allfour_RNAseq, blind=FALSE)
cus.plotPCA_34(dds_allfour_rl, intgroup = "condition_allfour")
cus.plotPCA_45(dds_allfour_rl, intgroup = "condition_allfour")
cus.plotPCA_12(dds_allfour_rl, intgroup = "condition_allfour")


#I will take the data has been batch effect corrected and rlog trasnformed and normalised by Deseq2

dds_allfour_rl_bf.pca <- FactoMineR::PCA(t(assay(dds_allfour_rl_bf)), graph = F, scale.unit = F)

#prepare for PCA plot layer
allfour_rl_bf_pca_ind <- as.data.frame(dds_allfour_rl_bf.pca$ind$coord) %>%
  dplyr::select(c(1:2))
allfour_rl_bf_pca_ind <- cbind(allfour_rl_bf_pca_ind, condition_allfour, batch_allfour)
allfour_rl_bf_pca_ind <- allfour_rl_bf_pca_ind %>%
  dplyr::rename("condition"="condition_allfour") %>%
  dplyr::rename("experiment"="batch_allfour")
allfour_rl_bf_pca_var_top20_name <- as.data.frame(dds_allfour_rl_bf.pca$var$contrib) %>%
  dplyr::slice_max(order_by = (Dim.1+Dim.2), n=20)
allfour_rl_bf_pca_var_top20_name <- row.names(allfour_rl_bf_pca_var_top20_name)
allfour_rl_bf_pca_var_top20 <- as.data.frame(dds_allfour_rl_bf.pca$var$coord) %>%
  tibble::rownames_to_column(var="Geneid") %>%
  dplyr::filter(Geneid %in% allfour_rl_bf_pca_var_top20_name) %>%
  dplyr::select(c(1:3)) %>%
  dplyr::inner_join(ensembl2name_loc, by=c("Geneid"="ensembl_gene_id")) %>%
  dplyr::mutate(X_start=0) %>%
  dplyr::mutate(Y_start=0)

#prepare the label of the x abd y axis in the plot. I modified from FactoMineR - plot.PCA source code in github: https://github.com/cran/FactoMineR/blob/master/R/plot.PCA.R
axes = c(1, 2)
lab.x <- paste("PC",axes[1], ": ",format(dds_allfour_rl_bf.pca$eig[axes[1],2],nsmall=0,digits=2),"% variance",sep="")
lab.y <- paste("PC ",axes[2], ": ",format(dds_allfour_rl_bf.pca$eig[axes[2],2],nsmall=0,digits=2),"% variance",sep="")
#use ggplot to plot
ggplot() + 
  geom_point(data=allfour_rl_bf_pca_ind, aes(Dim.1, Dim.2, col=condition, shape=experiment), size=5) + 
  scale_color_manual(values=c("#FE920A","#7F7F7F")) +
  geom_segment(data=allfour_rl_bf_pca_var_top20, aes(X_start, Y_start, xend = Dim.1*30, yend = Dim.2*30), arrow = arrow(length=unit(0.30,"cm"), ends="last", type = "closed"), color="#A569BD") +
  geom_text_repel(data=allfour_rl_bf_pca_var_top20, aes(Dim.1*30, Dim.2*30, label=external_gene_name), nudge_x=0.2, nudge_y = ifelse((allfour_rl_bf_pca_var_top20$Dim.2)>0, 0.6, -0.6), color="#A569BD", fontface=ifelse(allfour_rl_bf_pca_var_top20$external_gene_name %in% c("ATAD1", "PAPSS2", "PTEN"), "bold" ,"plain") )+
  xlab(lab.x) +
  ylab(lab.y) +
  theme_bw(base_size = 15) 

#adjust the label position if still overlap.
```

#volcano plot
```{r}
#volcano plot
allfour_HAP1_RNAseq_deseq2$padj[is.na(allfour_HAP1_RNAseq_deseq2$padj)] <- 1
#there are some gene whose -log10(padj) is too high, around 150, most of them are below 40. it will make figure very ugly. so I set -log10(pdj) above 40 as 40. And after generating the figure, I manually fix the y axis. change 40 to >=40.
allfour_HAP1_RNAseq_deseq2 <- allfour_HAP1_RNAseq_deseq2 %>%
  dplyr::mutate(padj_adj=(ifelse(padj>10^(-40), padj, 10^(-40))))
ggplot(allfour_HAP1_RNAseq_deseq2, aes(log2FoldChange,-log10(padj_adj), label=external_gene_name)) + 
  geom_point(color="grey", size=0.8) +
  geom_point(data = allfour_HAP1_RNAseq_deseq2 %>% filter(padj_adj<=0.05 & log2FoldChange>0), color="#B2182B", size=2) +
  geom_point(data = allfour_HAP1_RNAseq_deseq2 %>% filter(padj_adj<=0.05 & log2FoldChange<0), color="#0271C6", size=2) +
  geom_point(data = allfour_HAP1_RNAseq_deseq2 %>% filter(external_gene_name %in% c("PAPSS2", "ATAD1", "CFL1P1", "PTEN", "KLLN")), size=2, color="#0271C6") +
  geom_line(aes(y=-log10(0.05)), size =0.7, linetype = 2, color="black") +
  geom_line(aes(y=-log10(0.01)), size =0.7, linetype = 2, color="black") +
  theme_bw() +
 geom_text_repel(data = allfour_HAP1_RNAseq_deseq2 %>% filter(external_gene_name %in% c("PAPSS2", "ATAD1", "CFL1P1", "PTEN", "KLLN")), fontface = "italic", max.overlaps = Inf) +
  theme(axis.text = element_text(size = 20), 
        axis.title = element_text(size = 20),
        plot.title = element_text(size = 20, face = "bold", hjust = 0.5),
        plot.subtitle = element_text(size = 15, face = "bold", hjust = 0.5)) +
  xlab("log2(FoldChange)") +
  ylab("-log10(adj.P)")
#when run this, the error message: 	Removed >19k rows containing missing values (geom_point). It is because DEseq2 take all the count without filter so have lots of 0 read count gene, leads to log2FC and so on as NA. if type: sum(is.na(allfour_HAP1_RNAseq_deseq2$log2FoldChange)). the number is the same. so no need to worry.
#manually correct the y axis. from 40 to >=40
```


#DE gene distribution
```{r}
#DE genes distribution
#plot to see if the DE genes are enriched at certain chromosome.
chr <- read.delim("/Users/keyigeng/Dropbox/R_notebook/hg38_chrom_sizes.txt", header=F)
allfour_HAP1_RNAseq_DE_chrsum <- DEnumber_chrsize(allfour_HAP1_RNAseq_DEdeseq2)
 

#plot the number/freqeuncy of DE genes on each chromosome.
chr_order <- as.character(chr$V1)
chr_size <- allfour_HAP1_RNAseq_DE_chrsum$chr_size
ggplot(allfour_HAP1_RNAseq_DE_chrsum, aes(x=factor(chr, level=chr_order))) +
  geom_line(aes(y=DEnumber, group = 1), size = 1, color="#FEBA3F") + 
  geom_line(aes(y=DE_density*(10^8), group = 1), size = 1, color="#E07007") +  
  scale_y_continuous(name = "# DE genes",sec.axis = sec_axis(~./(10^8), name="# DE genes/chr length"))+
  xlab("") +
  theme_bw(base_size = 15) +
  theme(axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
        axis.text.y = element_text(size = 20 ),
        axis.text.y.right = element_text(size = 18 ),
        axis.title.y = element_text(color = "#FEBA3F", size=22),
        axis.title.y.right = element_text(color = "#E07007", size=22)) 

#plot the local changes (chr10) DE genes location.
allfour_deseq2_chr10 <- left_join(allfour_HAP1_RNAseq_deseq2, ensembl2name_loc, by=c("genes"= "ensembl_gene_id")) %>%
    distinct(genes, .keep_all = TRUE)%>%
  dplyr::filter(chromosome_name==10) %>%
  dplyr::select(!c("external_gene_name.x")) %>%
  dplyr::rename(external_gene_name=external_gene_name.y)
ggplot(allfour_deseq2_chr10, 
       aes(x=end_position, y=log2FoldChange)) +
  geom_point(size=0.8, color="grey", alpha=0.5) + 
  geom_line(aes(y=0), size =0.7, linetype = "dashed", color="black", alpha=1) +
  geom_point(data = allfour_deseq2_chr10 %>% filter( padj <= 0.05), color="#E69F00", size=1, alpha=1) +
  scale_x_continuous(labels=fancy_scientific) +
  ylab("log2FoldChange") +
  xlab("position") +
  theme(panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.line = element_line(colour = "black"), 
        panel.background = element_blank(), 
        aspect.ratio = 0.2, 
        axis.text = element_text(size = 15),
        axis.title = element_text(size = 15)) +
  geom_text_repel(
    data=allfour_deseq2_chr10 %>% dplyr::filter(padj <= 0.05) %>% dplyr::filter(abs(log2FoldChange) > 2) , aes(label=external_gene_name),
    nudge_x = 0.5, nudge_y = -1, min.segment.length = unit(0.1, "lines")
  ) + 
  geom_rect(aes(xmin = 87702682, xmax = 87985236, ymin = -Inf, ymax = Inf),
                   fill = "red")
#adjust the subscript and italic manually.
```


#GO, KEGG enrichment and visualization.
```{r}
#GO enrichment
allfour_RNAseq_DE_OR_GOBP_UP <-GO_enrichment_human((allfour_HAP1_RNAseq_DEdeseq2 %>% dplyr::filter(log2FoldChange>0)), "BP")
allfour_RNAseq_DE_OR_GOBP_DOWN <-GO_enrichment_human((allfour_HAP1_RNAseq_DEdeseq2 %>% dplyr::filter(log2FoldChange<0)), "BP")  
allfour_RNAseq_DE_OR_GOMF_UP <-GO_enrichment_human((allfour_HAP1_RNAseq_DEdeseq2 %>% dplyr::filter(log2FoldChange>0)), "MF") 
allfour_RNAseq_DE_OR_GOMF_DOWN <-GO_enrichment_human((allfour_HAP1_RNAseq_DEdeseq2 %>% dplyr::filter(log2FoldChange<0)), "MF") 

#GO cluster and visualization, cluster need to set seed if want to be reproducible. cluster without setting seed has also been tested a lot of time, conclusion is not changed, the 
#when the terms are not lots, the name of the cluster can be not very accurate, adjust it a bit.
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOBP_DOWN, y=30, z=1.5, m=0.7)
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOMF_DOWN, y=30, z=1.5, m=0.7)
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOBP_UP, y=30, z=1.5, m=0.7, nCluster=4)
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOMF_UP, y=26, z=1.5, m=0.7)
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOBP_DOWN, y=200, z=1, m=0.7, node_label="none") # allfour_RNAseq_DE_OR_GOBP_DOWN_pair_t only has 200 row even there are 358 enriched terms. Any number above 200, will receive error message.
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOMF_DOWN, y=66, z=1, m=0.7, node_label="none")
set.seed(123)
custome_emapplot(allfour_RNAseq_DE_OR_GOBP_UP, y=152, z=1, m=0.7, node_label="none")

#KEGG pathway
allfour_RNAseq_DE_OR_KEGG_DOWN <- KEGG_fun_human(allfour_HAP1_RNAseq_DEdeseq2  %>% filter (log2FoldChange<0)) 
allfour_RNAseq_DE_OR_KEGG_UP <- KEGG_fun_human(allfour_HAP1_RNAseq_DEdeseq2  %>% filter (log2FoldChange>0))
allfour_RNAseq_DE_OR_KEGG <- KEGG_fun_human(allfour_HAP1_RNAseq_DEdeseq2) 
allfour_RNAseq_DE_OR_KEGG_combi <- rbind(get_KEGG_info_2(allfour_RNAseq_DE_OR_KEGG_UP,"upregulated",30), get_KEGG_info_2(allfour_RNAseq_DE_OR_KEGG_DOWN, "downregulated",30))

#barplot of KEGG
allfour_RNAseq_DE_OR_KEGG_combidf <- allfour_RNAseq_DE_OR_KEGG_combi %>%
  dplyr::slice_min(n = 20, order_by=p.adjust) %>%
  dplyr::arrange(desc(p.adjust))
ggplot(allfour_RNAseq_DE_OR_KEGG_combidf, aes(x=Count, y=factor(Description, level=Description))) +
  geom_bar(aes(fill=-log10(p.adjust)), stat="identity") +
  scale_fill_gradient(high = "#F1BA1D", low = "#F8F4E9") +
  theme_bw() +
  facet_grid(~ trend) +
  theme(text=element_text(size = 15),
        plot.title = element_text(hjust = 1),
        strip.text.x = element_text(size = 9.5)) +
  xlab("") +
  ylab("") +
  labs(fill="-log10(p.adj)")

allfour_RNAseq_DE_OR_KEGG_DOWN_x <- setReadable(allfour_RNAseq_DE_OR_KEGG_DOWN, 'org.Hs.eg.db', 'ENTREZID')
allfour_RNAseq_DE_down_entrezid <- dplyr::inner_join((allfour_HAP1_RNAseq_DEdeseq2%>% dplyr::filter (log2FoldChange<0)), (bitr(allfour_HAP1_RNAseq_DEdeseq2$genes, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb=org.Hs.eg.db)), by=c("genes"="ENSEMBL"))
genelist_allfour_RNAseq_DE_down <- allfour_RNAseq_DE_down_entrezid$log2FoldChange
names(genelist_allfour_RNAseq_DE_down) <- allfour_RNAseq_DE_down_entrezid$ENTREZID
cnetplot(allfour_RNAseq_DE_OR_KEGG_DOWN_x, showCategory=5,categorySize="pvalue", foldChange = genelist_allfour_RNAseq_DE_down, circular=F ) +  scale_color_gradient(high = "#BFE2FE", low ="#0161AE")
```


#heatmap of the DE genes using DEseq2 normalised readscount.
```{r}
allfour_DE_deseqnor_count  <- as.data.frame(allfour_RNAseq_deseqnor_count) %>%
  tibble::rownames_to_column(var="gene") %>%
  dplyr::inner_join(allfour_HAP1_RNAseq_DEdeseq2, by=c("gene"="genes")) %>%
  dplyr::arrange(log2FoldChange) %>%
  dplyr::select(c(2,3,4,5,6,7,8,9,10,11,14,15,12,13,18,19,16,17)) %>%
  as.matrix()
allfour_DE_deseqnor_count_hm <- allfour_DE_deseqnor_count[,1:4]
allfour_DE_deseqnor_count_GSE103405 <- allfour_DE_deseqnor_count[,11:14]
allfour_DE_deseqnor_count_GSE147770 <- allfour_DE_deseqnor_count[,5:10]
allfour_DE_deseqnor_count_GSE148175 <- allfour_DE_deseqnor_count[, 15:18]
allfour_DE_deseqnor_count_list <- list(allfour_DE_deseqnor_count_hm, allfour_DE_deseqnor_count_GSE103405, allfour_DE_deseqnor_count_GSE147770, allfour_DE_deseqnor_count_GSE148175)
col_heatmap_2 <- colorRampPalette(c("#2166AC", "#4393C3","#92C5DE","#D1E5F0","#FFFFFF", "#FDDBC7", "#F4A582","#D6604D","#B2182B"))
allDE_deseqnor_count_sep <- lapply(allfour_DE_deseqnor_count_list, function(x) pheatmap(x, scale="row", cluster_cols=F, cluster_rows=F,cellwidth = 10, cellheight = 0.2,  breaks = seq(-2, 2, length.out = 30), border_color = "black", col= ( col_heatmap_2(30))))
names(allDE_deseqnor_count_sep) <- c("allfour_DE_deseqnor_count_hm", "allfour_DE_deseqnor_count_GSE103405", "allfour_DE_deseqnor_count_GSE147770", "allfour_DE_deseqnor_count_GSE148175")

```

#get the DE gene location to prepare for the bed files to calculate the distance between DB peaks and nearesr DE genes.
```{r}
allfour_HAP1_RNAseq_DE_loc <- allfour_HAP1_RNAseq_DEdeseq2 %>%
  dplyr::select(c(9,10,11,1))
allfour_HAP1_RNAseq_DE_loc$chromosome_name <- sub("^", "chr", allfour_HAP1_RNAseq_DE_loc$chromosome_name)
```



