---
title: "R Notebook"
output: html_notebook
---

#this is the R notebook for plotting the figures in PTEN-PAPSS2 deletion project. 
```{r}
library(data.table)
library(ggplot2)
library(stringr)
library(dplyr)
library(ggpubr)
library(rstatix)
library(ggstatsplot)
library(plyr)
library(clusterProfiler)
library(org.Hs.eg.db)
library("biomaRt")
library(tidyverse)
library(pheatmap)
library(ggfortify)
library(limma)
library("FactoMineR")
library("factoextra")
library(VennDiagram)
```


#functions
```{r}
###function to select pateint based on both copy number data and RNA-seq data.
#x is the input zsem data for each cancer type. y is the cutoff read count for considering it as del.
getdelID_fromzsem = function(x, y) {
  P_P_rsemdel <- x %>%
  dplyr::filter(Entrez_Gene_Id %in% P_P_entrzID) %>%
  tibble::column_to_rownames(var="Entrez_Gene_Id") %>%
  dplyr::select(!Hugo_Symbol)
P_P_rsemdel <- P_P_rsemdel[,colSums(P_P_rsemdel)<=y]
intersect(colnames(P_P_rsemdel), as.vector(del_patient_corID))
}

#x is the input zsem data for each cancer type. y is the cutoff read count for considering it as nor. z is the number identified with getdelID_fromzsem to get the matched normal.
#add skcm, ucec and cesc, now extract the patients without deletion.
getnorID_matchtype_fromzsem = function(x, y, z) {
  P_P_rsemdel <- x %>%
    dplyr::filter(Entrez_Gene_Id %in% P_P_entrzID) %>%
    tibble::column_to_rownames(var="Entrez_Gene_Id") %>%
    dplyr::select(!Hugo_Symbol)
  P_P_rsemdel <- P_P_rsemdel[ ,colSums(P_P_rsemdel)>=y] 
  sample(intersect(colnames(P_P_rsemdel), as.vector(nor_patient_corID)), z)
}

#extract the selected patient gene expression data and pre-filter the gene for the downstream analysis. PCA and heatmap.
#x is the input rsem file, y is the cutoff : 10x (sample number), a is the selected patient with del, b is the selected patient with intact region and decent expression.
get_exp_prefilter = function(x,y,a,b) {
  filterbyexp <- x[, which ((names(x) %in% c(a, b, "Entrez_Gene_Id"))==T), drop=F]
  filterbyexp <- filterbyexp[rowSums(filterbyexp[,-1])> y, ]
  filterbyexp$Entrez_Gene_Id <- as.character(filterbyexp$Entrez_Gene_Id)
  filterbyexp <- as.data.frame(filterbyexp) %>%
    drop_na() %>%
    group_by(Entrez_Gene_Id) %>%
    summarise_all(sum) %>%  #to sum up the row with the same gene ID
    tibble::remove_rownames() %>%
    tibble::column_to_rownames(var="Entrez_Gene_Id") 
  filterbyexp_PCA <- as.data.frame(t(filterbyexp)) %>%
    tibble::rownames_to_column(var = "sample") %>%
    dplyr::mutate(condition=ifelse(sample %in% a, "deletion", "with")) %>%
    tibble::column_to_rownames(var = "sample") %>%
    dplyr::arrange(factor(condition, levels=c("with", "deletion")))
  filterbyexp_PCA
}


heatmap_top100down=function(x, y) {
  col_heatmap_2 <- colorRampPalette(c("#2166AC", "#4393C3","#92C5DE","#D1E5F0","#FFFFFF", "#FDDBC7", "#F4A582","#D6604D","#B2182B"))
  pheatmap(log10(as.matrix(t(x[, which((names(x) %in% as.vector(HAP1DE_entrez_downtopx$entrezgene_id))==T)])+1)),  scale="row", cellwidth = 10, cluster_cols=F, cluster_rows=T, cellheight = 8, border_color =NA, col= ( col_heatmap_2(30)), annotation_col = x[, length(x), drop=F],  labels_row = (HAP1DE_entrez_downtopx[match(colnames(x[, which((names(x) %in% as.vector(HAP1DE_entrez_downtopx$entrezgene_id))==T)]), HAP1DE_entrez_downtopx$entrezgene_id), ])$external_gene_name.x,  annotation_colors=list(condition=c(with="#7F7F7F", deletion="#FE920A")), filename=y)
}

heatmap_topkegg = function(x, y) {col_heatmap_2 <- colorRampPalette(c("#2166AC", "#4393C3","#92C5DE","#D1E5F0","#FFFFFF", "#FDDBC7", "#F4A582","#D6604D","#B2182B"))
 pheatmap(log10(as.matrix(t(x[, which((names(x) %in% as.vector(genename_top5KEGG$entrezgene_id))==T)])+1)),  scale="row", cellwidth = 10, cluster_cols=F, cluster_rows=T, cellheight =9,border_color =NA, col= ( col_heatmap_2(30)), annotation_col = x[, length(x), drop=F],  labels_row = (genename_top5KEGG[match(colnames(x[, which((names(x) %in% as.vector(genename_top5KEGG$entrezgene_id))==T)]), genename_top5KEGG$entrezgene_id), ])$external_gene_name,  annotation_colors=list(condition=c(with="#7F7F7F", deletion="#FE920A")), 
filename=y) 
}

heatmap_DEDB=function(x, y) {
  col_heatmap_2 <- colorRampPalette(c("#2166AC", "#4393C3","#92C5DE","#D1E5F0","#FFFFFF", "#FDDBC7", "#F4A582","#D6604D","#B2182B"))
  pheatmap(log10(as.matrix(t(x[, which((names(x) %in% as.vector(DE_DB_5kb_entrez$entrezgene_id))==T)])+1)),  scale="row", cellwidth = 10, cluster_cols=F, cluster_rows=T, cellheight = 8, border_color =NA, col= ( col_heatmap_2(30)), annotation_col = x[, length(x), drop=F],  labels_row = (DE_DB_5kb_entrez[match(colnames(x[, which((names(x) %in% as.vector(DE_DB_5kb_entrez$entrezgene_id))==T)]), DE_DB_5kb_entrez$entrezgene_id), ])$external_gene_name.x,  annotation_colors=list(condition=c(with="#7F7F7F", deletion="#FE920A")), filename=y)
}


```


#the genomic alteration freqeuncy of PAPSS2, ATAD1, KLLN and PTEN accross different cancer types. input summary data is directly downloaded from cBioportal
```{r}

files=list.files(path="/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal", pattern="*_summary.txt", full.names=T)
temp <- lapply(files, data.table::fread)
genes_ROI_TCGA <- c("ATAD1", "KLLN","PAPSS2", "PTEN")
cBioPortal_TCGA_del_locus <- do.call(rbind, Map(cbind, temp, Gene=genes_ROI_TCGA)) 
rm(temp, files)
colnames(cBioPortal_TCGA_del_locus)<-str_replace_all(colnames(cBioPortal_TCGA_del_locus), c(" " = "_"))
ggplot(data=cBioPortal_TCGA_del_locus, aes(x=Cancer_Type, y=Alteration_Frequency, fill=Alteration_Type)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme_bw() +
  xlab("") +
  ylab("frequency (%)") +
  facet_grid(~factor(Gene, levels=c("PAPSS2", "ATAD1", "KLLN", "PTEN")), scales = "free")

```


#look into patient with deep deletion
```{r}

#read multiple files of individual patients' copy number data at specific gene locus (PAPSS2, ATAD1, KLLN, PTEN) into seperate data frames.
files_cna=list.files(path="/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/copy_number", pattern="cna_*")
for(i in files_cna){
    filepath <- file.path("/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/copy_number", i)
    assign(i, read.delim(filepath))
}

#find out patients with  4 genes co-deletion.
del_patient <- Reduce(intersect, list(cna_ATAD1.txt[cna_ATAD1.txt$ATAD1==-2, 2], cna_KLLN.txt[cna_KLLN.txt$KLLN==-2, 2], cna_PAPSS2.txt[cna_PAPSS2.txt$PAPSS2==-2, 2], cna_PTEN.txt[cna_PTEN.txt$PTEN==-2, 2])) #195 patient
nor_patient <- Reduce(intersect, list(cna_ATAD1.txt[cna_ATAD1.txt$ATAD1==0, 2], cna_KLLN.txt[cna_KLLN.txt$KLLN==0, 2], cna_PAPSS2.txt[cna_PAPSS2.txt$PAPSS2==0, 2], cna_PTEN.txt[cna_PTEN.txt$PTEN==0, 2])) 
del_patient_info <- cna_ATAD1.txt %>%
  dplyr::filter(SAMPLE_ID %in% del_patient) %>%
  dplyr::select(STUDY_ID, SAMPLE_ID)


#venn diagram
venn.diagram(
  x=list(cna_PAPSS2.txt[cna_PAPSS2.txt$PAPSS2==-2, 2], cna_ATAD1.txt[cna_ATAD1.txt$ATAD1==-2, 2], cna_KLLN.txt[cna_KLLN.txt$KLLN==-2, 2], cna_PTEN.txt[cna_PTEN.txt$PTEN==-2, 2]),
  category.names = c("PAPSS2", "ATAD1", "KLLN", "PTEN"),
  filename="/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/all_datasets/plots/venn_4pcgene_del.svg",
  output=T,
# Output features
  imagetype="svg" ,
  height = 800 , 
  width = 800 , 
  resolution = 300,
  compression = "lzw",
# Circles
  lwd = 1,
  lty = 'blank',
  fill = c("#94C882", "#94C882", "#94C882", "#94C882"),
# Numbers
  cex = 1,
  fontface = "bold",
  fontfamily = "sans",
# Set names
  cat.cex = 1,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans"
)

#to check which cancer type has enough patient as replicates to do analysis based on copy number variation data, cutoff set as 8.
del_patient_info_studyID_count <- del_patient_info %>%
  dplyr::count(STUDY_ID)
del_patient_info_studyID_count
del_patient_studyID <- as.vector((del_patient_info_studyID_count %>% dplyr::filter(n>8))[,1])

```



```{r}
#read in normalized RNA-seq read count on different cancer types. downloaded data format: folder named after the studyID, and the normalised RNA-seq data is called data_mrna_seq_v2_rsem.txt.
for (i in del_patient_studyID){
  assign(paste(i, "_rsem",sep=""), read.delim(file.path("/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/all_datasets/", paste(i, "data_mrna_seq_v2_rsem.txt", sep = "/"))))
}

#the rsem data has different format, TCGA.XX.XXXX.XX instead of TCGA-XX-XXXX-XX
del_patient_info$modiID <- c(str_replace_all(del_patient_info$SAMPLE_ID, "-", "."))
del_patient_corID <- c(as.vector(del_patient_info$modiID))

#see in each selected cancer sutdy, how many patient sample we selected from copy number data also have RNA-seq data.
patient_bothdata <- data.frame()
for (i in paste(del_patient_studyID, "_rsem", sep="")) {
  temp_count <- sum(names(get(i)) %in% as.vector(del_patient_corID), drop = FALSE) 
  temp_df <- c(i, temp_count)
  patient_bothdata <- rbind(patient_bothdata, temp_df)
  rm(temp_count, temp_df)
}
patient_bothdata


#add gene expression data into consideration of selecting PAPSS2-PTEN deletion patients as well.
P_P_entrzID <- c("9060", "84896", "100144748", "5728")
#the following cancer type study have enough petients passing the filter.
P_Pdel_rsemdel_skcm <- getdelID_fromzsem(skcm_tcga_pan_can_atlas_2018_rsem, 400) # 4
P_Pdel_rsemdel_prad <- getdelID_fromzsem(prad_tcga_pan_can_atlas_2018_rsem, 400) #6 samples
P_Pdel_rsemdel_ucec <- getdelID_fromzsem(ucec_tcga_pan_can_atlas_2018_rsem, 400) # 4
P_Pdel_rsemdel_lusc <- getdelID_fromzsem(lusc_tcga_pan_can_atlas_2018_rsem, 400) #3 samples.

#the following have too few patients, excluded later.
getdelID_fromzsem(brca_tcga_pan_can_atlas_2018_rsem, 400) #0 saples
getdelID_fromzsem(gbm_tcga_pan_can_atlas_2018_rsem, 400) #0
getdelID_fromzsem(coadread_tcga_pan_can_atlas_2018_rsem, 400) # 2
getdelID_fromzsem(stad_tcga_pan_can_atlas_2018_rsem, 400) # 0 

```


```{r}
#for the selected cancer type, select the patient in the matching cancer type with intact PAPSS2-PTEN locus with decent expression.
nor_patient_info <- cna_ATAD1.txt %>%
  dplyr::filter(SAMPLE_ID %in% nor_patient) %>%
  dplyr::select(STUDY_ID, SAMPLE_ID) %>%
  dplyr::filter(STUDY_ID %in% c("skcm_tcga_pan_can_atlas_2018", "prad_tcga_pan_can_atlas_2018", "ucec_tcga_pan_can_atlas_2018", "lusc_tcga_pan_can_atlas_2018"))
nor_patient_info$modiID <- c(str_replace_all(nor_patient_info$SAMPLE_ID, "-", "."))
nor_patient_corID <- c(as.vector(nor_patient_info$modiID))
```


###!!!IMPORTANT NOTE!!!!
#this chunk involves randomly take the patients with intact PAPSS2-PTEN locus to match the number of patients with deletion and the PCA plot. Therefore, it has been run >10 times to reduce the bias. In these 10 times run, the conclusion of PCA plot largely remains unchanged. PC1-PC3 tend to seperate the two groups.
#For PCA plots used in manuscripts and the downstream heatmap analysis, the patient sample ID used can be found in files deposit on github.
```{r}
P_Pnor_rsemdel_prad <- getnorID_matchtype_fromzsem(prad_tcga_pan_can_atlas_2018_rsem, 3000, 6)
P_Pnor_rsemdel_lusc <- getnorID_matchtype_fromzsem(lusc_tcga_pan_can_atlas_2018_rsem, 3000, 3)
P_Pnor_rsemdel_skcm <- getnorID_matchtype_fromzsem(skcm_tcga_pan_can_atlas_2018_rsem, 3000, 4)
P_Pnor_rsemdel_ucec <- getnorID_matchtype_fromzsem(ucec_tcga_pan_can_atlas_2018_rsem, 3000, 4)

resem_skcm_filterbyexp <- get_exp_prefilter(skcm_tcga_pan_can_atlas_2018_rsem, 80, P_Pdel_rsemdel_skcm, P_Pnor_rsemdel_skcm)
resem_prad_filterbyexp <- get_exp_prefilter(prad_tcga_pan_can_atlas_2018_rsem, 120, P_Pdel_rsemdel_prad, P_Pnor_rsemdel_prad)
resem_lusc_filterbyexp <- get_exp_prefilter(lusc_tcga_pan_can_atlas_2018_rsem, 60, P_Pdel_rsemdel_lusc, P_Pnor_rsemdel_lusc)
resem_ucec_filterbyexp <- get_exp_prefilter(ucec_tcga_pan_can_atlas_2018_rsem, 80, P_Pdel_rsemdel_ucec, P_Pnor_rsemdel_ucec)

autoplot(prcomp(as.matrix(log2(resem_skcm_filterbyexp[1:(length(resem_skcm_filterbyexp)-1)]+1)), center = T, scale = F) ,  data=resem_skcm_filterbyexp, colour="condition", x=1, y=3, size=2) + scale_color_manual(values=c("#FE920A","#7F7F7F")) + theme_bw(base_size = 20) 
autoplot(prcomp(as.matrix(log2(resem_prad_filterbyexp[1:(length(resem_prad_filterbyexp)-1)]+1)), center = T, scale = F) ,  data=resem_prad_filterbyexp, colour="condition", x=1, y=2, size=2) + scale_color_manual(values=c("#FE920A","#7F7F7F")) + theme_bw(base_size = 20) 
autoplot(prcomp(as.matrix(log2(resem_lusc_filterbyexp[1:(length(resem_lusc_filterbyexp)-1)]+1)), center = T, scale = F) ,  data=resem_lusc_filterbyexp, colour="condition", x=1, y=2, size=2) + scale_color_manual(values=c("#FE920A","#7F7F7F")) + theme_bw(base_size = 20) 
autoplot(prcomp(as.matrix(log2(resem_ucec_filterbyexp[1:(length(resem_ucec_filterbyexp)-1)]+1)), center = T, scale = F) ,  data=resem_ucec_filterbyexp, colour="condition", x=2, y=3, size=2) + scale_color_manual(values=c("#FE920A","#7F7F7F")) + theme_bw(base_size = 20) 

```

#plot the heatmap for subset of DE genes identified using HAP1 as model. sampleID used see in the documents.
```{r}
#Top 100 downregulated genes in HAP1.
mart <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl") 
ensembl2entrez <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id"),mart=mart)
#read the file containing DE genes identifiied from RNA-seq data.
allfour_HAP1_RNAseq_DEdeseq2 <- read.delim("/Users/keyigeng/Dropbox/HAP1_Gen_insta/files/allfour_HAP1_RNAseq_DEdeseq2.txt", sep=" ")
HAP1DE_entrez_downtopx <- allfour_HAP1_RNAseq_DEdeseq2 %>%
  dplyr::inner_join(ensembl2entrez, by=c("genes"="ensembl_gene_id")) %>%
  dplyr::slice_min(log2FoldChange, n=100) 

heatmap_top100down(resem_prad_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/prad_top100DEdown_log10.pdf") 
heatmap_top100down(resem_skcm_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/skcm_top100DEdown_log10.pdf")
heatmap_top100down(resem_lusc_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/lusc_top100DEdown_log10.pdf")
heatmap_top100down(resem_ucec_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/ucec_top100DEdown_log10.pdf")

#top 5 KEGG pathways.
#read the enriched top 5 KEGG pathway table from RNA-seq DE analysis
DE_OR_KEGG_DOWN <- read.delim ("/Users/keyigeng/Dropbox/HAP1_Gen_insta/files/allfour_RNAseq_DE_OR_KEGG_DOWN.txt")
topKEGG_HAP1_DE_genelist_topx <- DE_OR_KEGG_DOWN %>%
   dplyr::slice_min(p.adjust, n=5) %>%  
   dplyr::select(geneID)
topKEGG_HAP1_DE_genelist_topx <- unique(unlist(str_split(topKEGG_HAP1_DE_genelist_topx$geneID, "/")))
genename_top5KEGG <- ensembl2entrez %>%
  dplyr::filter(entrezgene_id %in% topKEGG_HAP1_DE_genelist_topx) %>%
  dplyr::select(!ensembl_gene_id) %>%
  dplyr::distinct() 
heatmap_topkegg(resem_prad_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/prad_topkegg_log10.pdf")
heatmap_topkegg(resem_lusc_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/lusc_topkegg_log10.pdf")
heatmap_topkegg(resem_skcm_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/skcm_topkegg_log10.pdf")
heatmap_topkegg(resem_ucec_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/ucec_topkegg_log10.pdf")

#DE-DB genes (DE genes with DB H3K27ac peak within 5kb)
#read the file containing the distance of each DE gene to nearest DB peak, and filter it out the gene with distance shorter than 5kb.
DEDB_5kb <- read.delim("/Users/keyigeng/Dropbox/HAP1_Gen_insta/ChIP-seq/files/Raul_DBpeak_20220228/DE_RaulDB_220228_dis_220228.txt", header=F) %>% 
  dplyr::filter(V9 <= 5000)
DE_DB_5kb_entrez <- ensembl2entrez %>%
  dplyr::filter(ensembl_gene_id %in% DEDB_5kb$V4) %>%
  drop_na() %>%
  dplyr::inner_join(allfour_HAP1_RNAseq_DEdeseq2, by=c("ensembl_gene_id"="genes")) %>%
  dplyr::arrange(log2FoldChange)
heatmap_DEDB(resem_prad_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/prad_DEDB_log10.pdf")
heatmap_DEDB(resem_lusc_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/lusc_DEDB_log10.pdf")
heatmap_DEDB(resem_skcm_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/skcm_DEDB_log10.pdf")
heatmap_DEDB(resem_ucec_filterbyexp, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/ucec_DEDB_log10.pdf")

```


```{r}
#write the file of the patient sample ID used for figure to enable reproducibility
sampleID_used_del <- str_replace_all(c(P_Pdel_rsemdel_lusc, P_Pdel_rsemdel_skcm, P_Pdel_rsemdel_prad, P_Pdel_rsemdel_ucec), "\\.", "-")
sampleID_used_del <- data.frame(sampleID_used_del) %>%
  dplyr::mutate(condition ="deletion") %>%
  dplyr::inner_join(cna_ATAD1.txt, by=c("sampleID_used_del"="SAMPLE_ID")) %>%
  dplyr::select(-4)
colnames(sampleID_used_del) <- c("SampleID", "Condition", "StudyID")

sampleID_used_nor <- str_replace_all(c(P_Pnor_rsemdel_lusc, P_Pnor_rsemdel_skcm, P_Pnor_rsemdel_prad, P_Pnor_rsemdel_ucec), "\\.", "-")
sampleID_used_nor <- data.frame(sampleID_used_nor) %>%
  dplyr::mutate(condition ="with") %>%
  dplyr::inner_join(cna_ATAD1.txt, by=c("sampleID_used_nor"="SAMPLE_ID")) %>%
  dplyr::select(-4)
colnames(sampleID_used_nor) <- c("SampleID", "Condition", "StudyID")

sampleID_used <- rbind(sampleID_used_nor, sampleID_used_del)
write.table(sampleID_used, "/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/plots/sampleID_used_info.txt", row.names = F, quote = F)
```


#look into the association between losing this locus with receiving radiation therapy 
```{r}
for (i in unique(c(del_patient_info$STUDY_ID))){
  assign(paste(i, "_patient_info",sep=""), read.delim(file.path("/Users/keyigeng/Dropbox/HAP1_Gen_insta/clinical_relevance/cBioPortal/all_datasets/", paste(i, "data_clinical_patient.txt", sep = "/"))))
}

nor_patient_info_2 <- cna_ATAD1.txt %>%
  dplyr::filter(SAMPLE_ID %in% nor_patient) %>%
  dplyr::select(STUDY_ID, SAMPLE_ID)

nor_patient_info_studyID_count <- nor_patient_info_2 %>%
  dplyr::count(STUDY_ID)
nor_patient_info_studyID_count

patient_info_nor <- data.frame()
for (i in unique(c(del_patient_info$STUDY_ID))){
  temp_patientID <- c(gsub('.{3}$', '', nor_patient_info_2$SAMPLE_ID)) #remove the last three character, which is the sample ID.
  temp_df <- get(paste(i, "_patient_info",sep="")) %>%
    dplyr::filter(X.Patient.Identifier %in% temp_patientID) %>%
    dplyr::mutate(source=i)
  patient_info_nor <- rbind(patient_info_nor, temp_df)
  rm(temp_patientID, temp_df)
}


patient_nor_radia <- patient_info_nor %>%
  dplyr::group_by(source) %>%
  dplyr::summarise(radia_yes_nor=sum(Radiation.Therapy=="Yes")) %>%
  dplyr::inner_join(nor_patient_info_studyID_count, by=c("source"="STUDY_ID")) 

patient_nor_radia_no <- patient_info_nor %>%
  dplyr::group_by(source) %>%
  dplyr::summarise(radia_no_nor=sum(Radiation.Therapy=="No")) 

patient_info_radia_deppdel_nor <- patient_nor_radia %>%
  dplyr::inner_join(patient_nor_radia_no, by=c("source"="source")) 
patient_info_radia_deppdel_nor
#for prad_tcga_pan_can_atlas_2018, the patient with copy number 0 at all four genes, with radiation therapy: 32, without radiation therapy: 260
#for the 6 patient samples selected above, with radiation therapy: 4, without radiation therapy: 2.

fisher.test(
  data.frame(
  "radia_no" = c(260, 2),  
  "radia_yes" = c(32, 4),
  row.names = c("0", "-2"),
  stringsAsFactors = FALSE
)
)
#p-value = 0.002287
prad_6samples_radia <- as.data.frame(cbind(c("with", "with", "deletion", "deletion"), c("radia_yes", "radia_no", "radia_yes", "radia_no"), c(32, 260, 4, 2)))
colnames(prad_6samples_radia) <- c("condition", "radia", "number")
prad_6samples_radia$number <- as.numeric(prad_6samples_radia$number)
prad_6samples_radia$condition <- factor(prad_6samples_radia$condition, levels=c("with", "deletion"))
ggplot(prad_6samples_radia, aes(fill = radia, y = number, x = condition)) +
  geom_bar(position = "fill", stat = "identity") + 
  labs(fill = "radiation", y = "fraction of clones", x="") + 
  theme_bw(base_size = 25) + 
  scale_fill_manual(values=c("#F0EBCE","#E0144C")) +
  theme(axis.text.x = element_text(size=20)) 


```




